React/NextJs/ProjectStructure.md

```diff
@@ -0,0 +1,212 @@
+# Project file structure
+
+## Organizing components
+(For PR Review):
+I believe this section will be refactored once we get around with all chakra stuff and component organization methodology.
+I kept going with only atomic design approach.
+
+### UI Components
+
+Before implementing [atomic design methodology](https://github.com/danilowoz/react-atomic-design), our common practice was separating components by the scope of the page + core components + shared components like this:
+
+Lowercase folders were indicators of a page scope (except of core).
+
+```
+.
+├── src
+└── components
+    ├── core
+    │   └── Button
+    │       ├── index.tsx
+    │       └── stories.tsx
+    ├── admin
+    │   └── WelcomeBanner
+    │       └── index.tsx
+    ├── Carousel
+    │   └── index.tsx
+    └── Card
+        └── index.tsx
+```
+
+With atomic design, we would completely separate all UI components and build them as a blocks of atoms, molecules and organisms (together with templates and pages for putting everything together at the end) 
+
+```
+src
+.
+└── components
+    └── ui
```

Comment:
Are ALL atoms, molecules and organisms under top-level `components/ui` dir, or do you allow for multiple `ui` dirs for multiple features? For example, some page might have some organism component that is used only on that page. It wouldn't make sense to place that organism under global `components/ui`, I would place it in some dir like `page/my-page/components/ui/organism`. Basically allow to have this structure that you explained here both globally and for feature-specific cases.

The reason why I am saying this is that if an app is really large, you will have 100+ organism in `components/ui/organism` and some of them might be used only on certain pages/features and some are real globals.

Reply:
@isBatak should probably give his feedback on this, but I believe we talked about the possibility of nesting components in such cases?
It would be good to explicitly mention it here.

Reply:
@DarkoKukovec Was there any idea of how to specifically nest such components, since they are not related to a context? 
@fvoska's example (`page/my-page/components/ui/organism`) also put's it into a context, which kind of beats the purpose ("At this level, the components begin to have the final shape, but they are still ensured to be independent, portable and reusable enough to be reusable in any content."), but in reality projects always get clogged and hard to read with a lot of files in a flat structure.

Reply:
@fvoska as quoted [here](https://github.com/infinum/JS-React-Example/blob/chakra/docs/atomic-design.md#organisms):

> Organisms are the combination of molecules that work together or even with atoms that compose more elaborate interfaces. At this level, the components begin to have the final shape, **but they are still ensured to be independent portable, and reusable enough to be reusable in any content**.

With that being said, organisms should still be reusable across different contexts/pages, and I believe their place is in the global ui folder.

I believe the mindset of context-bound (per page) organisms comes from one (or more) of the following situations:
- small project === small number of pages/features which results in a small number or organisms (no problem in traversing them at this point)
- component architecture is not setup as per the methodology violating the beforementioned principle (bolded in the quote)
- the design system is either bad and/or "versatile" resulting in the "inherent" creation of independent components. This might be out of scope here, but as projects get bigger there should be a consensus with the designers yadayada

Worst case scenario: you have a huge amount of code and you would like to scope the organisms in some manner; I'd propose to create subfolders which are scoped to either features or pages like so:

```
ui/
  organisms/
    feature1/
      some-organism
    feature2/
      some-organism
```

Reply:
@fvoska  Per page component namespacing is kinda a wrong mindset here.
Organisms are basically "container" components but in the new React ecosystem, you don't have to fetch data at the top-level anymore, instead, you can fetch data in any component and still be able to reuse that component wherever you want. We collocate data fetching with a component to make it highly reusable, therefore you don't have to model your component based on the place where it is rendered, instead, you can model it by the data you are presenting.

This is the definition of the Organism that I found on the internet  
> Organisms are the combination of molecules that work together or even with atoms that compose more elaborate interfaces. At this level, the components begin to have the final shape, but they are still ensured to be independent, portable, and reusable enough to be reusable in any content.

I copied the whole article here, maybe we can reuse some definitions for our handbook.
https://github.com/infinum/JS-React-Example/blob/chakra/docs/atomic-design.md#organisms

[Templates](https://github.com/infinum/JS-React-Example/blob/chakra/docs/atomic-design.md#templates) and a good design system (like Chakra UI), kinda resolve this problem and need to scope components to a specific page. They shape the organisms to fit on a specific page... 

In conclusion, all of this is really new to us and we need to take it slow. 
I wouldn't go with per page namespacing because it seems like a wrong abstraction.

Reply:
> React had a vision that goes beyond the browser. they understood that a clean component model is the unifying factor, that splitting up the view layer is a logical and technical mistake. the view and the code that forms it belong together, only state should be separate.